crates/ty_ide/src/symbols.rs

```diff
@@ -435,6 +517,234 @@ impl SymbolVisitor {
         symbol_id
     }
 
+    /// Adds a symbol introduced via an assignment.
+    fn add_assignment(&mut self, stmt: &ast::Stmt, name: &ast::ExprName) -> SymbolId {
+        let kind = if Self::is_constant_name(name.id.as_str()) {
+            SymbolKind::Constant
+        } else if self
+            .iter_symbol_stack()
+            .any(|s| s.kind == SymbolKind::Class)
+        {
+            SymbolKind::Field
+        } else {
+            SymbolKind::Variable
+        };
+
+        let symbol = SymbolTree {
+            parent: None,
+            name: name.id.to_string(),
+            kind,
+            name_range: name.range(),
+            full_range: stmt.range(),
+            import_kind: None,
+        };
+        self.add_symbol(symbol)
+    }
+
+    /// Adds a symbol introduced via an import `stmt`.
+    fn add_import_alias(&mut self, stmt: &ast::Stmt, alias: &ast::Alias) -> SymbolId {
+        let name = alias.asname.as_ref().unwrap_or(&alias.name);
+        let kind = if stmt.is_import_stmt() {
+            SymbolKind::Module
+        } else if Self::is_constant_name(name.as_str()) {
+            SymbolKind::Constant
+        } else {
+            SymbolKind::Variable
+        };
+        let re_export = Some(
+            if alias.asname.as_ref().map(ast::Identifier::as_str) == Some(alias.name.as_str()) {
+                ImportKind::RedundantAlias
+            } else {
+                ImportKind::Normal
+            },
+        );
+        self.add_symbol(SymbolTree {
+            parent: None,
+            name: name.id.to_string(),
+            kind,
+            name_range: name.range(),
+            full_range: stmt.range(),
+            import_kind: re_export,
+        })
+    }
+
+    /// Extracts `__all__` names from the given assignment.
+    ///
+    /// If the assignment isn't for `__all__`, then this is a no-op.
+    fn add_all_assignment(&mut self, targets: &[ast::Expr], value: Option<&ast::Expr>) {
+        if self.in_function || self.in_class {
+            return;
+        }
+        let Some(target) = targets.first() else {
+            return;
+        };
+        if !is_dunder_all(target) {
+            return;
+        }
+
+        let Some(value) = value else { return };
+        match *value {
+            // `__all__ = [...]`
+            // `__all__ = (...)`
+            ast::Expr::List(ast::ExprList { ref elts, .. })
+            | ast::Expr::Tuple(ast::ExprTuple { ref elts, .. }) => {
+                self.update_all_origin(DunderAllOrigin::CurrentModule);
+                if !self.add_all_names(elts) {
```

Comment:
It seems like in this case we should not treat `foo` as part of `__all__`? That is, direct (non-augmenting) assignment should replace any previous names.

```py
__all__ = ["foo"]
__all__ = ["bar"]
```

But I guess if we did this, then we'd also need to track whether an assignment to `__all__` was conditional, and understand statically-known conditions... so maybe the best option is to just collect everything we see and not try to understand any of that?

This constellation of limitations at least seems worth a comment, though, similar to the one about `submodule.__all__` below, because the ty version does model these things.